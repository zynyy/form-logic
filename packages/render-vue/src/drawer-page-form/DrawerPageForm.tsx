import { computed, defineComponent, ref, watch } from 'vue';

import DrawerModalFooter from '@/components/drawer-modal-footer';
import { SchemaPatternEnum } from '@/interface';
import SchemeForm from '@/scheme-form';
import PageLoading from '@/components/page-loading';
import { usePageForm } from '@/hooks';
import { useTriggerLogic } from '@/hooks/useLogic';
import { ClickRecord } from '@/components/buttons';
import { getDrawerPageFormProps, DrawerPageFormProps } from './interface';
import { getSubmitFormValues } from '@/utils/formUtils';

import { Drawer, message } from 'ant-design-vue';

const DrawerPageForm = defineComponent({
  name: 'DrawerPageForm',
  inheritAttrs: false,
  props: getDrawerPageFormProps(),
  setup(props: DrawerPageFormProps, { slots }) {
    const submitLoading = ref(false);

    const loadingDone = ref(false);

    const [triggerLogic] = useTriggerLogic(props.getLogicConfig, () => {
      submitLoading.value = false;
    });

    const setSubmitLoading = (nextSubmitLoading: boolean) => {
      submitLoading.value = nextSubmitLoading;
    };

    const pageFormArgsRef = computed(() => {
      const { formConfig, onFormMount, getLogicConfig, events, extraLogicParams, options } = props;

      return {
        formConfig,
        onFormMount,
        options,
        getLogicConfig,
        logicParams: extraLogicParams,
        events,
        autoRefreshForm: false,
      };
    });

    const pageFormRef = usePageForm(pageFormArgsRef);

    const submit = (formValues: any) => {
      const { onConfirm } = props;

      onConfirm?.(formValues);

      submitLoading.value = false;
      handleCloseClick();
    };

    const handleConfirmClick = () => {
      const { form } = pageFormRef.value;

      const { validateFormValues } = props;
      setSubmitLoading(true);
      getSubmitFormValues(form)
        .then((formValues) => {
          if (validateFormValues) {
            validateFormValues(formValues)
              .then(() => {
                submit(formValues);
              })
              .catch((err) => {
                setSubmitLoading(false);
                if (err) {
                  message.warning(err).then(() => void 0);
                }
              });
          } else {
            submit(formValues);
          }
        })
        .catch(() => {
          setSubmitLoading(false);
        });
    };

    const handleCloseClick = (e?: MouseEvent) => {
      const { onClose } = props;

      onClose?.(e);
    };

    const handleAfterVisibleChange = (visibleDone) => {
      loadingDone.value = visibleDone;
    };

    const handleButtonItemClick = (e, record: ClickRecord) => {
      const { eventCode, clickCodes, code } = record;

      const { events, extraLogicParams, options } = props;

      const { form } = pageFormRef.value;

      if (events?.[eventCode]) {
        return events[eventCode](e, {
          form,
          setSubmitLoading,
          code,
        });
      }

      if (clickCodes.length) {
        setSubmitLoading(true);

        triggerLogic(clickCodes, {
          params: extraLogicParams,
          form,
          effectHook: 'onClick',
          fieldCode: code,
          pageCode: options?.metaSchema?.code,
        });
      }
    };

    const renderFooter = () => {
      const { buttons } = pageFormRef.value;

      const { components, hasConfirmButton, options } = props;

      return (
        <DrawerModalFooter
          buttons={buttons}
          hasConfirmButton={hasConfirmButton ?? options?.pattern !== SchemaPatternEnum.DETAIL}
          onClick={handleButtonItemClick}
          onConfirmClick={handleConfirmClick}
          onCloseCLick={handleCloseClick}
          loading={submitLoading.value}
          disabled={submitLoading.value}
          components={components}
        />
      );
    };

    const renderTitle = () => {
      if (slots.title) {
        return slots.title();
      }

      const { title } = props;

      return title || '请填写';
    };

    watch(
      () => {
        return {
          formConfig: props.formConfig,
          visible: props.visible,
        };
      },
      () => {
        setSubmitLoading(false);
        if (props.visible) {
          const { refreshForm } = pageFormRef.value;
          refreshForm(props.formConfig);
        }
      },
    );

    return () => {
      const {
        components,

        width,
        extraLogicParams,
        getLogicConfig,
        events,
        visible,
        language,
      } = props;

      const { schema, form, pattern, formLoading } = pageFormRef.value;

      const loading = formLoading || !visible || !loadingDone.value;

      return (
        <Drawer
          maskClosable={false}
          visible={visible}
          onClose={handleCloseClick}
          closable={false}
          width={width || '90%'}
          v-slots={{
            title: renderTitle,
            footer: renderFooter,
          }}
          destroyOnClose
          onAfterVisibleChange={handleAfterVisibleChange}
        >
          <PageLoading loading={submitLoading.value}>
            <SchemeForm
              form={form}
              loading={loading}
              schema={schema}
              components={components}
              pattern={pattern}
              getLogicConfig={getLogicConfig}
              events={events}
              extraLogicParams={extraLogicParams}
              language={language}
            />
          </PageLoading>
        </Drawer>
      );
    };
  },
});

export default DrawerPageForm;
