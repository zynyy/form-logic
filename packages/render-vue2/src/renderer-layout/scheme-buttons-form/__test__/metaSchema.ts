import { MetaSchema } from '@/interface';

export default {
  code: 'Api_C',
  labelCol: 6,
  wrapperCol: 18,
  defaultSearchColumn: 6,
  group: [],
  data: [
    {
      type: 'column',
      schemaType: 'string',
      required: '0',
      hidden: '0',
      disabled: '0',
      hiddenName: '0',
      wrap: '0',
      logics: [],
      code: 'url',
      name: '请求接口',
      hasSerialNo: '0',
      hasSort: '0',
    },
    {
      type: 'column',
      schemaType: 'string',
      required: '1',
      hidden: '0',
      disabled: '0',
      hiddenName: '0',
      wrap: '0',
      logics: [],
      code: 'method',
      name: '请求方式',
      defaultValue: 'get',
      component: 'input',
    },
  ],
  model: 'Api',
  codeSuffix: 'C',
  name: '组件接口配置',
} as MetaSchema;
