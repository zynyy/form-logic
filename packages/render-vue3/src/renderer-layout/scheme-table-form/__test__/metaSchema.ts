import { MetaSchema } from '@/interface';

export default {
  code: 'Logic_L',
  name: '逻辑列表',
  model: 'Logic',
  labelCol: 6,
  wrapperCol: 18,
  columnLayout: 3,
  group: [],
  defaultSearchColumn: 6,
  data: [
    {
      code: 'code',
      name: '编码',
      type: 'search_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'name',
      name: '名称',
      type: 'search_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'addBtn',
      name: '新建',
      type: 'button',
      schemaType: 'void',
      component: 'ArrayBase.Addition',
      logics: [],
    },
    {
      code: 'code',
      name: '编码',
      type: 'table_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'name',
      name: '名称',
      type: 'table_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'remarks',
      name: '备注说明',
      type: 'table_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'editBtn',
      name: '编辑',
      type: 'table_button',
      schemaType: 'void',
      component: 'ArrayBase.Edit',
      logics: [],
    },
    {
      code: 'removeBtn',
      name: '删除',
      type: 'table_button',
      schemaType: 'void',
      component: 'ArrayBase.Remove',
      logics: [],
    },
    {
      type: 'table_button',
      schemaType: 'string',
      required: '0',
      hidden: '0',
      disabled: '0',
      logics: [],
      code: 'detailBtn',
      name: '详情',
      component: 'ArrayBase.Detail',
    },
  ],
} as MetaSchema;
