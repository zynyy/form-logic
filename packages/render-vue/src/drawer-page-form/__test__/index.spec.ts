import { mount } from '@vue/test-utils';

import DrawerPageForm from '../DrawerPageForm';

import metaSchema from './metaSchema';

describe('drawer-page-form ', () => {
  let wrapper;
  let drawerPageForm;

  beforeEach(() => {
    wrapper = mount(DrawerPageForm, {
      props: {
        visible: true,
        options: {
          metaSchema,
          hasGroup: true,
        },
        getLogicConfig: () => {
          return Promise.resolve();
        },
        formConfig: {},
      },
    });

    drawerPageForm = wrapper.getComponent(DrawerPageForm);
  });

  test('gen snapshot', async () => {
    expect(document.body.outerHTML).toMatchSnapshot();
  });
});
