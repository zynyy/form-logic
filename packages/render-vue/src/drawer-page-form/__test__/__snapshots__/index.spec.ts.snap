// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawer-page-form  gen snapshot 1`] = `
<body style="overflow: hidden; overflow-x: hidden; overflow-y: hidden;"
      class="ant-scrolling-effect"
>
  <div>
    <div tabindex="-1"
         class="ant-drawer ant-drawer-right ant-drawer-open"
    >
      <div class="ant-drawer-mask">
      </div>
      <div class="ant-drawer-content-wrapper"
           style="width: 90%;"
      >
        <div class="ant-drawer-content">
          <div class="ant-drawer-wrapper-body">
            <div class="ant-drawer-header">
              <div class="ant-drawer-header-title">
                <div class="ant-drawer-title">
                  请填写
                </div>
              </div>
            </div>
            <div class="ant-drawer-body">
              <div class="ant-spin-nested-loading form-logic-page-loading">
                <div class="ant-spin-container">
                  <div class="ant-skeleton">
                    <div class="ant-skeleton-content">
                      <h3 class="ant-skeleton-title"
                          style="width: 38%;"
                      >
                      </h3>
                      <ul class="ant-skeleton-paragraph">
                        <li>
                        </li>
                        <li>
                        </li>
                        <li style="width: 61%;">
                        </li>
                      </ul>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div class="ant-drawer-footer">
              <div class="ant-row form-logic-left-right-slot">
                <div class="ant-col ant-col-12 form-logic-left-right-slot-left">
                  <div class="ant-space ant-space-horizontal ant-space-align-center">
                    <div class="ant-space-item">
                      <button class="ant-btn ant-btn-dashed"
                              type="button"
                      >
                        <span role="img"
                              aria-label="close"
                              class="anticon anticon-close"
                        >
                          <svg focusable="false"
                               class
                               data-icon="close"
                               width="1em"
                               height="1em"
                               fill="currentColor"
                               aria-hidden="true"
                               viewbox="64 64 896 896"
                          >
                            <path d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z">
                            </path>
                          </svg>
                        </span>
                        <span>
                          关闭
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
                <div class="ant-col ant-col-12 form-logic-left-right-slot-right">
                  <div class="ant-space ant-space-horizontal ant-space-align-center">
                    <div class="ant-space-item"
                         style="margin-right: 8px;"
                    >
                    </div>
                    <div class="ant-space-item">
                      <button class="ant-btn ant-btn-primary"
                              type="button"
                      >
                        <span role="img"
                              aria-label="check"
                              class="anticon anticon-check"
                        >
                          <svg focusable="false"
                               class
                               data-icon="check"
                               width="1em"
                               height="1em"
                               fill="currentColor"
                               aria-hidden="true"
                               viewbox="64 64 896 896"
                          >
                            <path d="M912 190h-69.9c-9.8 0-19.1 4.5-25.1 12.2L404.7 724.5 207 474a32 32 0 00-25.1-12.2H112c-6.7 0-10.4 7.7-6.3 12.9l273.9 347c12.8 16.2 37.4 16.2 50.3 0l488.4-618.9c4.1-5.1.4-12.8-6.3-12.8z">
                            </path>
                          </svg>
                        </span>
                        <span>
                          确定
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
